pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -176,8 +169,51 @@ func (a *AmbientIndex) All() []*model.WorkloadInfo {
 	return res
 }
 
-func (c *Controller) AmbientSnapshot() *model.AmbientSnapshot {
-	return c.ambientIndex.ToSnapshot()
+func (c *Controller) WorkloadsForWaypoint(scope model.WaypointScope) []*model.WorkloadInfo {
+	a := c.ambientIndex
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	var res []*model.WorkloadInfo
+	// TODO: try to precompute
+	for _, w := range a.byPod {
+		if a.matchesScope(scope, w) {
+			res = append(res, w)
+		}
+	}
+	return res
+}
+
+func (c *Controller) Waypoint(scope model.WaypointScope) sets.Set[netip.Addr] {
+	a := c.ambientIndex
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	res := sets.New[netip.Addr]()
+	for ip := range a.waypoints[scope] {
+		res.Insert(netip.MustParseAddr(ip))
+	}
+
+	return res
+}
+
+func (a *AmbientIndex) matchesScope(scope model.WaypointScope, w *model.WorkloadInfo) bool {
+	if len(scope.ServiceAccount) == 0 {
+		// We are a namespace wide waypoint. SA scope take precedence.
+		// Check if there is one for this workloads service account
+		if _, f := a.waypoints[model.WaypointScope{Namespace: scope.Namespace, ServiceAccount: w.ServiceAccount}]; f {
+			return false
+		}
+	}
+	if scope.ServiceAccount != "" && (w.ServiceAccount != scope.ServiceAccount) {
+		return false
+	}
+	if w.Namespace != scope.Namespace {
+		return false
+	}
+	// Filter out waypoints.
+	if w.Labels[constants.ManagedGatewayLabel] == constants.ManagedGatewayMeshControllerLabel {
```

Comment:
gateway.istio.io/managed"= "istio.io-mesh-controller is the label for waypoint right?
Then why filter it out

Reply:
Waypoints do not have waypoint proxies. Note: this code is just moved